maro/rl/training/async_tools/actor.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import time
+from os import getcwd
+from typing import Callable, List, Union
+
+from maro.communication import Proxy, SessionMessage
+from maro.rl.wrappers import AbsEnvWrapper, AgentWrapper
+from maro.utils import Logger
+
+from ..message_enums import MsgKey, MsgTag
+
+
+def actor(
+    create_env_wrapper_func: Callable[[], AbsEnvWrapper],
+    create_agent_wrapper_func: Callable[[], AgentWrapper],
+    num_episodes: int,
+    group: str,
+    num_steps: int = -1,
+    create_eval_env_wrapper_func: Callable[[], AbsEnvWrapper] = None,
+    eval_schedule: Union[int, List[int]] = None,
+    log_env_summary: bool = True,
+    proxy_kwargs: dict = {},
+    log_dir: str = getcwd()
+):
+    """Controller for single-threaded learning workflows.
+
+    Args:
+        create_env_wrapper_func (Callable): Function to create an environment wrapper for training data collection.
+            The function should take no parameters and return an environment wrapper instance.
+        create_agent_wrapper_func (Callable): Function to create a decision generator for interacting with
+            the environment. The function should take no parameters and return a ``AgentWrapper`` instance.
+        num_episodes (int): Number of training episodes. Each training episode may contain one or more
+            collect-update cycles, depending on how the implementation of the roll-out manager.
+        num_steps (int): Number of environment steps to roll out in each call to ``collect``. Defaults to -1, in which
+            case the roll-out will be executed until the end of the environment.
+        create_env_wrapper_func (Callable): Function to create an environment wrapper for evaluation. The function
+            should take no parameters and return an environment wrapper instance. If this is None, the training
+            environment wrapper will be used for evaluation. Defaults to None.
+        eval_schedule (Union[int, List[int]]): Evaluation schedule. If an integer is provided, the policies will
+            will be evaluated every ``eval_schedule`` episodes. If a list is provided, the policies will be evaluated
+            at the end of the training episodes given in the list. In any case, the policies will be evaluated
+            at the end of the last training episode. Defaults to None, in which case the policies will only be
+            evaluated after the last training episode.
+        log_env_summary (bool): If True, the ``summary`` property of the environment wrapper will be logged at the end
+            of each episode. Defaults to True.
+        proxy_kwargs: Keyword parameters for the internal ``Proxy`` instance. See ``Proxy`` class
+            for details. Defaults to the empty dictionary.
+        log_dir (str): Directory to store logs in. A ``Logger`` with tag "LOCAL_ROLLOUT_MANAGER" will be created at init
```

Comment:
Could this name configurable? I'm wondering whether there could be several `actor` at the same time?

Reply:
how about this one?

Reply:
I added actor_idx which can be read from the environment variables (line 32).